main.cpp

```diff
@@ -41,12 +41,207 @@
+// fuzzing of the bolt12 decoding logic without the fuzzer getting stuck on
+// fuzzing the bech32 decoding logic. This custom mutator is originally from
+// core-lightning:
+// https://github.com/ElementsProject/lightning/blob/release-v25.02.2/tests/fuzz/bolt12.h#L54
+extern "C" size_t LLVMFuzzerMutate(uint8_t *Data, size_t Size, size_t MaxSize);
+extern "C" size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+                                          unsigned int seed);
+extern "C" size_t LLVMFuzzerCustomCrossOver(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                                            size_t in2_size, uint8_t *out, size_t max_out_size,
+                                            unsigned seed);
+
+extern const std::string bech32_hrp;
+
+/* Encodes a dummy bolt12 offer/invoice-request/invoice into fuzz_data and
+ * returns the size of the encoded string. */
+static size_t initial_input(uint8_t *fuzz_data, size_t max_size)
+{
+    size_t output_size = std::min(max_size, dummy_initial_input.size());
+    std::memcpy(fuzz_data, dummy_initial_input.data(), output_size);
+
+    return output_size;
+}
+
+/* A custom mutator that decodes the bech32 input, mutates the decoded input,
+ * and then re-encodes the mutated input. This produces an input corpus that
+ * consists entirely of correctly encoded bech32 strings, enabling efficient
+ * fuzzing of the bolt12 decoding logic without the fuzzer getting stuck on
+ * fuzzing the bech32 decoding logic. */
+extern "C" size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size,
+                                          size_t max_size, unsigned int seed)
+{
+    try
+    {
+        if (size < 9)
+            return initial_input(fuzz_data, max_size);
+        std::string input_str(reinterpret_cast<char *>(fuzz_data), size);
+
+        // Check if input contains the correct HRP before decoding
+        if (input_str.length() < bech32_hrp.length() + 1 ||
+            input_str.substr(0, bech32_hrp.length()) != bech32_hrp ||
+            (input_str.length() > bech32_hrp.length() && input_str[bech32_hrp.length()] != '1'))
+        {
+            return initial_input(fuzz_data, max_size);
+        }
+        
+        // Decode the input
+        bech32::DecodeResult decoded_result = bech32::DecodeNoChecksum(input_str, bech32::CharLimit::CUSTOM_MUTATOR);
+        std::vector<uint8_t> decoded_data = decoded_result.data;
+
+        if (decoded_data.empty())
+        {
+            return initial_input(fuzz_data, max_size);
+        }
+
+        if (decoded_data.size() > max_size)
+        {
+            return initial_input(fuzz_data, max_size);
+        }
+
+
+        // Mutate the data part of the decoded input
+        std::vector<uint8_t> mutated_data = decoded_data;
+
+        // Resize the vector to the maximum possible size BEFORE mutation.
+        mutated_data.resize(std::max(mutated_data.size(), max_size));
```

Comment:
Shouldn't we just truncate to `max_size` if necessary?

```suggestion
        mutated_data.resize(max_size));
```